### PR TITLE
62 docs generation fsharp formatting

### LINKED ISSUE #62: Docs Generation
## Description

Like project scaffold, it would be nice to have document generation. Need to research more into this.

## Error messages, screenshots

n/a

## Failing test, failing github repo, or reproduction steps

n/a

## Expected Behavior

n/a

## Known workarounds

/na

### PR DESCRIPTION
**Pull request by TheAngryByrd**
## Proposed Changes

Adding Documentation Generation 

Closes #62 

## Types of changes

What types of changes does your code introduce to MiniScaffold?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Build and tests pass locally
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by TheAngryByrd** [comment]
Thanks to @ominestre for a lot of the conversion help!

**Comment by TheAngryByrd** [comment]
@baronfel I have hot reloading for things in `docsSrc`. Know of a way we can do it for the `fsx` files under templates since they are needed to be reloaded into the currently running fake script?

**Comment by baronfel** [comment]
I pushed up some fixes for comments, we had the rendering of them inside an if-expression that was a bit too broad. To clarify the structure of the `partMembers` function I broke it out into named pieces.

**Comment by TheAngryByrd** [comment]
Things left to do:

- [ ] Generating a Navigation structure from the folder structure
- [ ] Good styling / structure of pages 
- [ ] Dealing with different Target Frameworks (should we care about this?, this isn't a showstopper to me)
- [ ] Dealing with multiple (older) versions of the documentation (could be put into a separate issue, this isn't a showstopper to me)

@baronfel @jdodds I think we need to just build a Trie for the nav structure.  Then building that into a sub nav dropdown structure with bootstrap. 

@andrewpucci I think we're at the point where I'll need your help whenever you get a chance to save me from my design skills.

File: Content/Library/src/MyLib.1/paket.references [hunk 1]
```diff
@@ -1,2 +1,3 @@
 FSharp.Core
 Microsoft.SourceLink.GitHub
+Newtonsoft.Json
```
Comment (TheAngryByrd) [inline Content/Library/src/MyLib.1/paket.references]: Need to remove this before merge.  It was here to prove we could load 3rd party types in the docs

File: Content/Library/src/MyLib.1/Library.fs [hunk 2]
```diff
@@ -1,6 +1,50 @@
...
+        sprintf "Hello %s" name
+
+    /// We did a bad api design
+    [<Obsolete>]
+    let reallyOldCode name =
+        nothing name
+
+
+    /// Who doesn't like json?
+    let personJToken (person : Person) =
```
Comment (TheAngryByrd) [inline Content/Library/src/MyLib.1/Library.fs]: Need to remove this before merge. It was here to prove we could load 3rd party types in the docs

File: Content/Library/src/MyLib.1/Library.fs [hunk 3]
```diff
@@ -1,6 +1,50 @@
...
+        name |> ignore
+
+
+    [<CompiledName("Hiya")>]
+    let hello name =
+        sprintf "Hello %s" name
+
+    /// We did a bad api design
+    [<Obsolete>]
+    let reallyOldCode name =
```
Comment (TheAngryByrd) [inline Content/Library/src/MyLib.1/Library.fs]: Need to remove this before merge. It was here to prove we show obsolete.

File: Content/Library/src/MyLib.1/Library.fs [hunk 4]
```diff
@@ -1,6 +1,50 @@
...
+            person.Last
+            (person.DateOfBirth.ToString("o"))
+            person.FavoriteNumber
+
+    /// I do nothing
+    let nothing name =
+        name |> ignore
+
+
+    [<CompiledName("Hiya")>]
```
Comment (TheAngryByrd) [inline Content/Library/src/MyLib.1/Library.fs]: Need to remove this before merge. It was here to prove we show CompiledName..

File: Content/Library/docsSrc/content/style.css [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+h3 a, h3 a:hover, h3 a:focus,
+h4 a, h4 a:hover, h4 a:focus,
+h5 a, h5 a:hover, h5 a:focus,
+h6 a, h6 a:hover, h6 a:focus { color : inherit; text-decoration : inherit; outline:none }
+
+/*--------------------------------------------------------------------------
+  Additional formatting for the homepage
+/*--------------------------------------------------------------------------*/
+
+#nuget {
```
Comment (andrewpucci) [inline Content/Library/docsSrc/content/style.css]: Don't think I'm seeing this id generated anywhere in the HTML. Am I missing something?

File: Content/Library/docsTool/templates/master.fs [hunk 6]
```diff
@@ -22,31 +22,31 @@ let renderFooter (cfg : MasterTemplateConfig) =
                         h5 [] [ str "Project Resources"]
                         ul [Class "list-group list-group-flush"] [
                             li [Class "list-group-item list-group-item-dark ml-0 pl-0"] [
-                                a [Href (cfg.GitHubRepoUrl |> Uri.simpleCombine "blob/master/README.md"); Class "text-white"] [
+                                a [Href (cfg.GitHubRepoUrl |> Uri.simpleCombine "blob/master/README.md"); Target "_blank"; Class "text-white"] [
```
Comment (andrewpucci) [inline Content/Library/docsTool/templates/master.fs]: https://css-tricks.com/use-target_blank/#article-header-id-2

File: Content/Library/docsSrc/content/style.css (hunk 5 above)
Comment (TheAngryByrd) [reply Content/Library/docsSrc/content/style.css]: I just pulled the base from https://github.com/fsprojects/FSharp.Formatting/blob/master/docs/files/content/style.css so anything we dont use we can just purge.

**Comment by TheAngryByrd** [comment]
This is pretty feature complete for a v1.  

@baronfel @ominestre @jdodds any reviews would be appreciated before I do my last rounds.

File: Content/Library/src/MyLib.1/Library.fs [hunk 7]
```diff
@@ -1,6 +1,74 @@
...
+    let hello name =
+        sprintf "Hello %s" name
+
+    /// We did a bad api design
+    [<Obsolete>]
+    let reallyOldCode (name : 'a) =
```
Comment (TheAngryByrd) [inline Content/Library/src/MyLib.1/Library.fs]: Need to remove before merge.  It was here to prove we show Obsolete

File: Content/Library/docsTool/templates/partNested.fs [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+                            td [] [
+                                Helpers.createAnchorIcon t.Name
+                            ]
+                            td [Class "type-name"] [
+                                a [Href (sprintf "%s.html" t.UrlName)] [
+                                    str t.Name
+                                ]
+                            ]
+                            td [Class "xmldoc"] [
+                                if t.IsObsolete then
```
Comment (baronfel) [inline Content/Library/docsTool/templates/partNested.fs]: I've seen this obsolete chunk in a few places, is it worth it to extract it for uniform representation?

File: Content/Library/docsTool/templates/types.fs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                        yield str (attr.Format())
+                        yield br []
+                ]
+        ]
+
+        yield div [
+            Class "xmldoc"
+        ] [
+            for sec in comment.Sections do
+                if byCategory |> Seq.exists (fun m -> m.GroupKey = sec.Key) |> not then
```
Comment (baronfel) [inline Content/Library/docsTool/templates/types.fs]: similarly, I've seen this xmldoc section repeated as well. seems like a good place for a component/function.

File: Content/Library/docsTool/templates/master.fs [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+    ProjectName : string
+    ReleaseVersion : string
+    ReleaseDate : DateTimeOffset
+}
+
+let renderFooter (cfg : MasterTemplateConfig) (pageSource : string option) =
+    footer [Class "footer font-small m-0 py-4 bg-dark"] [
+        div [Class "container"] [
+            div [Class "row"] [
+                div [Class "col-12 col-md-4 mb-4 mb-md-0"] [
```
Comment (baronfel) [inline Content/Library/docsTool/templates/master.fs]: these link-section divs feel like they could be a component as well, ie pull out the repo-local `a`-generation to a component, then the specific list-group-item formatting, then the overall header-with-list-of-items thing, etc.

**Comment by baronfel** [comment]
Few things:

## header

* should the menus animate on hover, not just click?

## footer
* contributing link isn't shown. Should these links not render if the backing file doesn't exist? `https://github.com/TheAngryByrd/miniscaffold-docs-test/blob/master/CONTRIBUTING.md`
* same for code of conduct
* The FSSF slack link won't work if you don't have a login, maybe it should point to `https://fsharp.org/guides/slack/` instead, which has registration instructions?
* 'edit this page' link can go right to the edit experience if we point to `/edit/` instead of `/blob/` in the url
* copyright notice doesn't have a date in it, should it be generated with one?

## landing page

* it feels like the api-docs button isn't anchored the same as the others. not quite in the bottom-left corner of the box?  and should they all be bottom-right instead? image
* 'learn usecases' button 404's, currently going to `https://www.jimmybyrd.me/miniscaffold-docs-test/HowTos/Doing_A_Thing.html`. is that the right url?

## Api reference

### type-level pages

* the `namespace` isn't rendered the same as it is in the module-level. There it's a bolded-heading with text underneath, here it's all plain text.
* For unions, do the docs include the field names if the union case fields have names?

### type signatures
* should names of types that are in this library be hyperlinked?

**Pull request merged**